// botbuilder-facebook-quick-replies/index.js
/**
 * @fileoverview Provides an easy way to add Facebook Quick Replies to a
 * message from bot to user with just one line of code.
 * @author Anton Ivanov <anton@ivanov.hk>
 */

/**
 * Provides an easy way to add Facebook Quick Replies to a
 * message from bot to user with just one line of code.
 */
const quickReplies = {
  /**
   * Adds the facebook quick replies to the builder.Message object.
   * Usage:
   *   let msg = new builder.Message(session).text('What do you think?');
   *   quick.replies(msg, ['Cool!', 'Not bad!']);
   *   session.send(msg);
   * @param {string} message - The botbuilder.Message object.
   * @param {string[]|object[]} texts - An array of strings, each of which
   *  will become a button. If an array of objects, each object must have a
   *  `text` and `value`.
   */
  replies: function (message, texts) {
    message.sourceEvent({
      facebook: {
        quick_replies: generateQuickReplies(texts),
      },
    });
  },
};

/**
 * Converts an array of strings into an array of quick_reply objects.
 * @param {string[]|object[]} texts
 * @returns {object[]} - An object array that can be added to quick_replies
 *    inside sourceEvent => facebook.
 */
function generateQuickReplies(texts) {
  let quick_replies = [];
  if (!texts || texts.length === 0) { return quick_replies; }

  for (let button of texts) {
    let text = typeof button === 'string' ? button : button.text;
    let payload = typeof button === 'string' ? button : button.value;

    quick_replies.push({
      content_type: 'text',
      title: text, // display value
      payload: payload, // actual value sent to the bot
      // image_url: 'https://...'
    });
  }

  return quick_replies;
}

module.exports = quickReplies;
